File: test/parity/canonical.c
/* test/parity/canonical.c -- see canonical.h for the contract. */
#include "canonical.h"

#include <inttypes.h>
#include <stdio.h>

const char THERMALCORE_CANONICAL_HEADER[] =
    "ts_ms,row_type,id,value,flags_or_status,a1,a2,a3,a4\n";

int thermalcore_canonical_sample(char *buf, size_t cap,
                                 uint32_t ts_ms, uint16_t signal_id,
                                 int32_t value, uint16_t flags)
{
    int n = snprintf(buf, cap,
                     "%" PRIu32 ",S,%u,%" PRId32 ",%u,0,0,0,0\n",
                     ts_ms, (unsigned)signal_id, value,
                     (unsigned)flags);
    return (n > 0 && (size_t)n < cap) ? n : 0;
}

int thermalcore_canonical_event(char *buf, size_t cap,
                                uint32_t ts_ms, uint16_t code,
                                uint32_t a1, uint32_t a2,
                                uint32_t a3, uint32_t a4)
{
    int n = snprintf(buf, cap,
                     "%" PRIu32 ",E,%u,0,0,%" PRIu32 ",%" PRIu32
                     ",%" PRIu32 ",%" PRIu32 "\n",
                     ts_ms, (unsigned)code, a1, a2, a3, a4);
    return (n > 0 && (size_t)n < cap) ? n : 0;
}

File: test/parity/canonical.py
"""test/parity/canonical.py -- Python mirror of test/parity/canonical.c.

The canonical telemetry CSV projection used by Stage 15.  This
module must stay byte-for-byte identical to canonical.c: the host
UDP probe (Stage 12 determinism) formats through here, the C
replay rigs (Stage 15 host-vs-target parity) format through
canonical.c, and both must hash the same projection.

Row layout, one row per core callback:

    ts_ms,row_type,id,value,flags_or_status,a1,a2,a3,a4

  - `S` (sample) rows -- id = signal_id, value = value,
    flags_or_status = flags, a1..a4 = 0.
  - `E` (event) rows  -- id = event_code, value = 0,
    flags_or_status = 0, a1..a4 = the four event args.

Decimal integers only, no whitespace, '\\n' line endings.
"""

CANONICAL_HEADER = "ts_ms,row_type,id,value,flags_or_status,a1,a2,a3,a4\n"


def canonical_sample(ts_ms, signal_id, value, flags=0):
    """Format one telemetry sample as a canonical `S` row."""
    return f"{ts_ms},S,{signal_id},{value},{flags},0,0,0,0\n"


def canonical_event(ts_ms, code, a1, a2, a3, a4):
    """Format one event as a canonical `E` row."""
    return f"{ts_ms},E,{code},0,0,{a1},{a2},{a3},{a4}\n"
